src/app-layer-pop3.c

```diff
@@ -0,0 +1,1080 @@
+
+            } else {
+                state->current_line = state->input;
+                state->current_line_len = lf_idx - state->input;
+
+                if (state->input != lf_idx &&
+                        *(lf_idx - 1) == 0x0D) {
+                    state->current_line_len--;
+                    state->current_line_delimiter_len = 2;
+                } else {
+                    state->current_line_delimiter_len = 1;
+                }
+            }
+
+            state->input_len -= (lf_idx - state->input) + 1;
+            state->input = (lf_idx + 1);
+
+            return 0;
+        } /* else - if (lf_idx == NULL) */
+    }
+
+}
+
+static int POP3InsertCommandIntoCommandBuffer(uint8_t command, POP3State *state, Flow *f)
+{
+    SCEnter();
+    void *ptmp;
+
+    if (state->cmds_cnt >= state->cmds_buffer_len) {
+        int increment = POP3_COMMAND_BUFFER_STEPS;
+        if ((int)(state->cmds_buffer_len + POP3_COMMAND_BUFFER_STEPS) > (int)USHRT_MAX) {
+            increment = USHRT_MAX - state->cmds_buffer_len;
+        }
+
+        ptmp = SCRealloc(state->cmds,
+                         sizeof(uint8_t) * (state->cmds_buffer_len + increment));
+        if (ptmp == NULL) {
+            SCFree(state->cmds);
+            state->cmds = NULL;
+            SCLogDebug("SCRealloc failure");
+            return -1;
+        }
+        state->cmds = ptmp;
+
+        state->cmds_buffer_len += increment;
+    }
+    if (state->cmds_cnt >= 1) {
+        /* decoder event */
+        AppLayerDecoderEventsSetEvent(f,
+                POP3_DECODER_EVENT_INVALID_PIPELINED_SEQUENCE);
+        /* we have to have EHLO, DATA, VRFY, EXPN, TURN, QUIT, NOOP,
+         * STARTTLS as the last command in pipelined mode */
+    }
+
+    /** \todo decoder event */
+    if ((int)(state->cmds_cnt + 1) > (int)USHRT_MAX) {
+        SCLogDebug("command buffer overflow");
+        return -1;
+    }
+
+    state->cmds[state->cmds_cnt] = command;
+    state->cmds_cnt++;
+
+    return 0;
+}
+
+static int POP3ProcessCommandRETR(POP3State *state, Flow *f,
+        AppLayerParserState *pstate)
+{
+    SCEnter();
+    SCReturnInt(0);
+}
+
+/* toclient */
+static int POP3ProcessReply(POP3State *state, Flow *f,
+        AppLayerParserState *pstate)
+{
+    SCEnter();
+
+//printf("current_command: %d\n", state->current_command);
+    switch (state->current_command) {
+        case POP3_COMMAND_RETR:
+
+            if (state->parser_state & POP3_PARSER_STATE_COMMAND_DATA_MODE) {
+            if (state->current_line_len == 1 && state->current_line[0] == '.') {
+                state->parser_state &= ~POP3_PARSER_STATE_COMMAND_DATA_MODE;
+
+                if (pop3_config.decode_mime &&
+                    (state->curr_tx != NULL) &&
+                    (state->curr_tx->mime_state != NULL)) {
+                    /* Complete parsing task */
+//printf("Mime parse complete\n");
+                    int ret = MimeDecParseComplete(state->curr_tx->mime_state);
```

Comment:
Can we unify handling of this between the email protocols instead of this duplication?